README.md
# Cana

![License](https://img.shields.io/hexpm/l/cana) ![Hex.pm](https://img.shields.io/hexpm/v/cana)

Cana is a simple email address validator based on RFC3696, RFC5322, RFC5321,
Wikipedia and various other reading.

## Project goals

Cana is a small, single purpose library, only for email address validation.

## Installation

Add cana to your mix.exs.

```elixir
def deps do
  [
    {:cana, "~> 0.1.0"}
  ]
end
```

## Examples
```elixir
iex> Cana.validate_email_address("justin@progressiveaspect.com")
{:ok, %{domain: "progressiveaspect.com", local: "justin"}}

iex> Cana.validate_email_address("badad\"dress@badplace.com")
{:error, "expected string \"@\" at column: 5"}
```

## Benchmarks

Taken with a teaspoon of salt; test of 10000 items.
```
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-4771 CPU @ 3.50GHz
Number of Available Cores: 2
Available memory: 7.75 GB
Elixir 1.10.4
Erlang 23.0.3

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 5 s
memory time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 14 s

Benchmarking invalid...
Benchmarking valid...

Name              ips        average  deviation         median         99th %
invalid         19.62       50.97 ms    ±13.79%       48.41 ms       91.28 ms
valid           10.76       92.94 ms    ±11.02%       88.66 ms      123.64 ms

Comparison: 
invalid         19.62
valid           10.76 - 1.82x slower +41.97 ms
```

## License
Copyright (c) 2020 Justin McPherson

This project is licensed under the terms of the MIT license, please see the
LICENCE.md file for more details.

CHANGELOG.md
### 02/10/2020

Release

test/cana_test.exs
defmodule CanaTest do
  use ExUnit.Case
  doctest Cana

  @valid_email_addresses [
    # https://en.wikipedia.org/wiki/Email_address#Examples
    {~S(simple@example.com), "simple", "example.com"},
    {~S(very.common@example.com), "very.common", "example.com"},
    {~S(disposable.style.email.with+symbol@example.com), "disposable.style.email.with+symbol",
     "example.com"},
    {~S(other.email-with-hyphen@example.com), "other.email-with-hyphen", "example.com"},
    {~S(fully-qualified-domain@example.com), "fully-qualified-domain", "example.com"},
    {~S(user.name+tag+sorting@example.com), "user.name+tag+sorting", "example.com"},
    {~S(x@example.com), "x", "example.com"},
    {~S(example-indeed@strange-example.com), "example-indeed", "strange-example.com"},
    {~S(admin@mailserver1), "admin", "mailserver1"},
    {~S(example@s.example), "example", "s.example"},
    {~S(" "@example.org), " ", "example.org"},
    {~S("john..doe"@example.org), "john..doe", "example.org"},
    {~S(mailhost!username@example.org), "mailhost!username", "example.org"},
    {~S(user%example.com@example.org), "user%example.com", "example.org"},

    # RFC3696
    {~S("Abc@def"@example.com), ~S(Abc@def), "example.com"},
    {~S(Abc\@def@example.com), ~S(Abc@def), "example.com"},
    {~S(Fred\ Blogs@example.com), ~S(Fred Blogs), "example.com"},
    {~S(Joe.\\Blow@example.com), ~S(Joe.\Blow), "example.com"},
    {~S(customer/department=shipping@example.com), "customer/department=shipping", "example.com"},
    {~S($A12345@example.com), "$A12345", "example.com"},
    {~S(!def!xyz%abc@example.com), "!def!xyz%abc", "example.com"},
    {~S(_somename@example.com), "_somename", "example.com"},

    # https://gist.github.com/cjaoude/fd9910626629b53c4d25
    {~S(email@example.com), "email", "example.com"},
    {~S(firstname.lastname@example.com), "firstname.lastname", "example.com"},
    {~S(email@subdomain.example.com), "email", "subdomain.example.com"},
    {~S(firstname+lastname@example.com), "firstname+lastname", "example.com"},
    {~S(email@123.123.123.123), "email", "123.123.123.123"},
    {~S(email@[123.123.123.123]), "email", "123.123.123.123"},
    {~S("email"@example.com), "email", "example.com"},
    {~S(1234567890@example.com), "1234567890", "example.com"},
    {~S(email@example-one.com), "email", "example-one.com"},
    {~S(_______@example.com), "_______", "example.com"},
    {~S(email@example.name), "email", "example.name"},
    {~S(email@example.museum), "email", "example.museum"},
    {~S(email@example.co.jp), "email", "example.co.jp"},
    {~S(firstname-lastname@example.com), "firstname-lastname", "example.com"}
    ## Seems these three are incorrect: the "dot-atom" and "quoted-string"
    ## forms are exclusive; RFC5322
    # {~S(much.”more\ unusual”@example.com), ~S(much.”more\ unusual”), "example.com"},
    # {~S(very.unusual.”@”.unusual.com@example.com), ~S(very.unusual.”@”.unusual.com), "example.com"},
    # {~S(very.”(),:;<>[]”.VERY.”very@\\ "very”.unusual@strange.example.com), ~S(very.”(),:;<>[]”.VERY.”very@\\ "very”.unusual@), "strange.example.com"}
  ]

  @invalid_email_addresses [
    # https://en.wikipedia.org/wiki/Email_address#Examples
    ~S(Abc.example.com),
    ~S(A@b@c@example.com),
    ~S(a"b(c\)d,e:f;g<h>i[j\k]l@example.com),
    ~S(just"not"right@example.com),
    ~S(this is"not\allowed@example.com),
    ## Seems this is incorrect: space, quote and backslash can be quoted by
    ## the backslash character; RFC3696
    # ~S(this\ still\"not\\allowed@example.com),
    ~S(1234567890123456789012345678901234567890123456789012345678901234+x@example.com),
    ~S(i_like_underscore@but_its_not_allow_in_this_part.example.com),

    # RFC3696
    ~S(john..doe@example.com),

    # https://gist.github.com/cjaoude/fd9910626629b53c4d25
    ~S(plainaddress),
    ~S(#@%^%#$@#$@#.com),
    ~S(@example.com),
    ~S(Joe Smith <email@example.com>),
    ~S(email.example.com),
    ~S(email@example@example.com),
    ~S(.email@example.com),
    ~S(email.@example.com),
    ~S(email..email@example.com),
    ~S(あいうえお@example.com),
    ~S[email@example.com (Joe Smith)],
    ## Out of scope; domain resolution is addressed in RFC5321, some
    ## suggestions on TLD verification in RFC3696
    # ~S(email@example),
    # ~S(email@example.web),
    ~S(email@-example.com),
    ~S(email@111.222.333.44444),
    ~S(email@example..com),
    ~S(Abc..123@example.com),
    ~S/”(),:;<>[\]@example.com/,
    ~S(just”not”right@example.com),
    ~S(this\ is"really"not\allowed@example.com),

    # Cana
    ~S(.john@example.com),
    ~S(john.@example.com),
    ~S(.john.@example.com)
  ]

  ##

  describe "ranges" do
    test "email address must be at least 6 characters" do
      assert {:error, "email address length (5) is outside of range 6..320"} =
               Cana.validate_email_address("a@b.c")
    end

    test "email address must be no more than 320 characters" do
      assert {:error, "email address length (321) is outside of range 6..320"} =
               Cana.validate_email_address(
                 ~s(#{String.duplicate("a", 63)}@#{String.duplicate("b", 63)}.#{
                   String.duplicate("c", 63)
                 }.#{String.duplicate("d", 63)}.#{String.duplicate("e", 62)}.au)
               )
    end

    test "local part must be no more than 64 characters" do
      assert {:error, "local part length(65) is outside the range 1..64"} =
               Cana.validate_email_address(~s(#{String.duplicate("a", 65)}@b.au))
    end

    test "domain part, labels must be no longer than 63 characters" do
      assert {:error, "domain label may not exceed 63 characters"} =
               Cana.validate_email_address(~s(a@#{String.duplicate("b", 64)}.au))
    end

    test "domain part must be no more than 255 characters" do
      assert {:error, "domain part length(256) is outside the range 4..255"} =
               Cana.validate_email_address(
                 ~s(#{String.duplicate("a", 63)}@#{String.duplicate("b", 63)}.#{
                   String.duplicate("c", 63)
                 }.#{String.duplicate("d", 63)}.#{String.duplicate("e", 61)}.au)
               )
    end
  end

  describe "examples" do
    test "success on valid email address" do
      @valid_email_addresses
      |> Enum.each(fn {address, local, domain} ->
        assert {:ok, %{local: ^local, domain: ^domain}} = Cana.validate_email_address(address)
      end)
    end

    test "failure on invalid address" do
      @invalid_email_addresses
      |> Enum.each(fn address ->
        assert {:error, _} = Cana.validate_email_address(address)
      end)
    end
  end

  describe "benchmarks" do
    @tag benchmark: true
    test "performance" do
      Benchee.run(%{
        "valid" => fn ->
          @valid_email_addresses
          |> Stream.cycle()
          |> Stream.take(10000)
          |> Enum.each(fn {address, _, _} ->
            Cana.validate_email_address(address)
          end)
        end,
        "invalid" => fn ->
          @invalid_email_addresses
          |> Stream.cycle()
          |> Stream.take(10000)
          |> Enum.each(fn address ->
            Cana.validate_email_address(address)
          end)
        end
      })
    end
  end
end


mix.exs
defmodule Cana.MixProject do
  use Mix.Project

  def project do
    [
      app: :cana,
      version: "0.1.0",
      elixir: "~> 1.10",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      package: package(),
      source_url: "https://github.com/justinmcp/cana"
    ]
  end

  def application do
    []
  end

  defp deps do
    [
      {:nimble_parsec, "~> 0.6.0", only: :dev},
      {:mix_test_watch, "~> 1.0", only: :dev, runtime: false},
      {:benchee, "~> 1.0", only: :test}
    ]
  end

  defp description() do
    "Simple email address validator"
  end

  defp package() do
    [
      files: ~w(lib/**/*.ex mix.exs README* LICENSE* CHANGELOG*),
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/justinmcp/cana"}
    ]
  end
end


lib/cana/validator.ex.exs
defmodule Cana.Validator do
  # parsec:Cana.Validator
  import NimbleParsec

  free_quote = utf8_char(empty(), [?"])
  escaped_quote = string(~S/\"/) |> replace("\"")
  escaped_space = string(~S/\ /) |> replace(" ")
  escaped_backslash = string(~S/\\/) |> replace("\\")
  escaped_at = string(~S/\@/) |> replace("@")
  at = string("@")
  dot = string(".")
  open_square = string("[")
  close_square = string("]")

  quoted_local =
    ignore(free_quote)
    |> utf8_string(
      [?a..?z, ?A..?Z, ?0..?9, ?\s, ?!, ?#..?/, ?=, ??, ?^, ?_, ?`, ?{, ?|, ?}, ?~, ?@],
      min: 1,
      max: 64
    )
    |> ignore(free_quote)

  unquoted_local =
    empty()
    |> utf8_string([not: ?.], 1)
    |> repeat(
      lookahead_not(string(".."))
      |> lookahead_not(string(".@"))
      |> choice([
        utf8_string([?a..?z, ?A..?Z, ?0..?9, ?!, ?#..?-, ?/, ?=, ??, ?^, ?_, ?`, ?{, ?|, ?}, ?~],
          min: 1,
          max: 64
        ),
        dot,
        escaped_quote,
        escaped_space,
        escaped_backslash,
        escaped_at
      ])
    )
    |> reduce({Enum, :join, [""]})

  local_part =
    choice([
      quoted_local,
      unquoted_local
    ])

  ip_address =
    optional(ignore(open_square))
    |> times(
      empty()
      |> utf8_string([?0..?9], max: 3)
      |> ignore(dot),
      3
    )
    |> utf8_string([?0..?9], max: 3)
    |> optional(ignore(close_square))
    |> reduce({Enum, :join, ["."]})

  domain_name =
    repeat(
      empty()
      |> lookahead_not(string("-"))
      |> utf8_string([?a..?z, ?A..?Z, ?0..?9, ?-], min: 1)
      |> optional(ignore(dot))
    )

  domain_part =
    choice([
      ip_address,
      domain_name
    ])

  defparsec :validate_email_address,
            local_part
            |> concat(at)
            |> concat(domain_part)

  # parsec:Cana.Validator
end


lib/cana.ex
defmodule Cana do
  @spec validate_email_address(String.t(), Keyword.t()) :: {:ok, map()} | {:error, String.t()}
  @doc ~S"""
  Validate an email address.

  Validate `email_address` according to RFC3696.

  ## Examples
      iex> Cana.validate_email_address("justin@progressiveaspect.com")
      {:ok, %{domain: "progressiveaspect.com", local: "justin"}}

      iex> Cana.validate_email_address("badad\"dress@badplace.com")
      {:error, "expected string \"@\" at column: 5"}
  """
  def validate_email_address(email_address, opts \\ []) when is_binary(email_address) do
    do_validate_email_address(email_address, opts, String.length(email_address))
  end

  defp do_validate_email_address(email_address, opts, length)
       when length >= 6 and length <= 320 do
    email_address
    |> Cana.Validator.validate_email_address(opts)
    |> handle_result()
  end

  defp do_validate_email_address(_email_address, _opts, length),
    do: {:error, "email address length (#{length}) is outside of range 6..320"}

  defp handle_result({:ok, [local, "@" | domain], "", _, _, _}) do
    %{
      local: local,
      domain: domain,
      error: nil
    }
    |> validate_local_length(String.length(local))
    |> validate_domain()
    |> convert_result()
  end

  defp handle_result({:ok, _, extra, _, _, column}) do
    {:error, "Unexpected #{extra} at column: #{column}"}
  end

  defp handle_result({:error, reason, _, _, _, column}) do
    {:error, "#{reason} at column: #{column}"}
  end

  defp validate_local_length(result, length) when length > 64 do
    %{result | error: "local part length(#{length}) is outside the range 1..64"}
  end

  defp validate_local_length(result, _), do: result

  defp validate_domain(%{domain: domain} = result) do
    if Enum.any?(domain, &(String.length(&1) > 63)) do
      %{result | error: "domain label may not exceed 63 characters"}
    else
      domain_part = Enum.join(domain, ".")
      validate_domain_length(%{result | domain: domain_part}, String.length(domain_part))
    end
  end

  defp validate_domain_length(result, length) when length > 255 do
    %{result | error: "domain part length(#{length}) is outside the range 4..255"}
  end

  defp validate_domain_length(result, _), do: result

  defp convert_result(%{error: error}) when not is_nil(error) do
    {:error, error}
  end

  defp convert_result(result) do
    {:ok, Map.drop(result, [:error])}
  end

  @type t :: __MODULE__
end
